[issues/131] - Added fork documentation (#134)
paritytech/foundry-polkadot: README.md

![Demo](.github/assets/demo.gif)

---

## Differences Between Foundry and Foundry-Polkadot Fork

This section outlines the key differences between the original Foundry toolkit ([Foundry](https://github.com/foundry-rs/foundry)) and the ParityTech Foundry-Polkadot fork ([Foundry-Polkadot](https://github.com/paritytech/foundry-polkadot)). **It is intended for developers already familiar with Foundry who are interested in using the fork for Polkadot blockchain development.** The focus is on new features, changes, and unsupported functionalities in the fork, emphasizing its adaptation for Polkadot.

### 1. Resolc Compiler Integration

> **Purpose**: The most significant addition in the fork is support for compiling Solidity contracts for Polkadot using the Resolc compiler, which targets PolkaVM in the `pallet-revive` framework ([Revive Repository](https://github.com/paritytech/revive)). This contrasts with the original Foundry's use of the Solc compiler for Ethereum's EVM.

Reviewer: The main addition in the foundry-polkadot is support for compiling Solidity contracts targeting Polkadot, using the Resolc compiler to generate PolkaVM-compatible bytecode. This contrasts with the original Foundry's use of the Solc compiler for Ethereum's EVM.

**Usage**:
- **Command-Line Flag**: For commands that involve compilation (e.g., `forge build`), you can use the `--resolc` flag to enable Resolc compilation. For example:
  ```bash
  forge build --resolc
  ```
  This instructs Forge to use Resolc instead of Solc, generating bytecode compatible with PolkaVM.
- **Configuration**: Alternatively, you can configure Resolc usage in the `foundry.toml` file by adding:
  ```toml
  [profile.default.resolc]
  resolc_compile = true
  ```

>   Setting `resolc_compile = false` reverts to using Solc, ensuring compatibility with Ethereum projects. By default, Foundry-Polkadot may use Solc unless Resolc is explicitly enabled, though this depends on the Foundry-Polkadot's configuration defaults.

Reviewer: ```suggestion
Setting `resolc_compile = false` reverts to using Solc, ensuring compatibility with Ethereum projects. By default, Foundry-Polkadot uses Solc unless Resolc is explicitly enabled
```

>   Setting `resolc_compile = false` reverts to using Solc, ensuring compatibility with Ethereum projects. By default, the fork may use Solc unless Resolc is explicitly enabled, though this depends on the fork's configuration defaults.

Reviewer: people will not understand what fork is, please put there foundry-polkadot or just foundry

Author: Addressed.

### 2. Command-Line Interface

> **Subcommands**: The fork Foundry currently supports a wide range of subcommands under `forge` ([Forge Commands](./README.forge.md)) and `cast` ([Cast Commands](./README.cast.md)).

Reviewer: again foundry polkadot

Author: Addressed.

> **Potential Differences**: Commands relying on compilation (e.g., `build`, `test`, `script`) in the fork can use Resolc, altering their output compared to the original Foundry.

Author: `forge test --resolc` gives fails. Addressed.

### 3. Unsupported (yet) or Modified Features

> **Potential Unsupported Features**: Some Ethereum-specific features, such as some Ethereum network interactions via Anvil and Chisel, may not be currently fully supported or relevant in the Polkadot context.

Reviewer: Support for Anvil and Chisel is not available, which also means commands like forge test, forge snapshot, and forge script are not currently supported.

> **Potential Unsupported Features**: Some Ethereum-specific features, such as EVM version targeting (e.g., `--evm-version`) or some Ethereum network interactions via Anvil and Chisel, may not be currently fully supported or relevant in the Polkadot context.

Reviewer: anvil chisel is not supported, evm version is accpeted and probobly used by solc when compiling to yul - to check

Author: Addressed.

**Modified Features**:

> - Compilation output: When using Resolc, the bytecode may differ from Solc's output, reflecting PolkaVM's requirements.

Reviewer: Resolc bytecode differs from Solc

Author: Addressed.

> - Compilation output: When using Resolc, the bytecode and ABI may differ from Solc's output, reflecting PolkaVM's requirements.

Reviewer: bytecode differs, please check ABI compatibility

Author: `forge inspect --json Counter abi` gives indeed the same output.

Reviewer: Please remember that you nned to use resolc: `forge inspect --resolc`

> - Project paths: Foundry-Polkadot adjusts project paths when `resolc_compile = true`, potentially affecting how source files, libraries, or artifacts are organized.

Author: Removing this since it's better described under the section `Resolc Compiler Integration`.
